### PR TITLE
LPS-177115: The get{customObjectName}PermissionsPage should return the actual Owner permissions

### PR DESCRIPTION
**Pull request by sergiojimcos**
**What is new?**
- Update `base_resource.ftl` implementing two new methods:

1. `_getResourcePermissionSet`: This function returns all the `ResourcePermissions` scoped by `GROUP`, `COMPANY`, `INDIVIDUAL`, `GROUP_TEMPLATE`.
2. `_populateResourcePermissionActionIds`: This function fills all the `actionsIds` given a Resource object and the scope they use.
- Execute `gw buildREST` in all `modules/apps` and `modules/apps/dxp`. 

**JIRA TICKET**
https://issues.liferay.com/browse/LPS-177115

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 1]
```diff
@@ -975,68 +977,114 @@ public abstract class Base${schemaName}ResourceImpl
 		}
 	}
 
-	private Collection<Permission> _getPermissions(long companyId, List<ResourceAction> resourceActions, long resourceId, String resourceName, String[] roleNames) throws Exception {
-		_checkResources(companyId, resourceId, resourceName);
+	private Collection<Permission> _getPermissions(
```
Comment (4lejandrito) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:980]: Can we autogenerate tests for this to ensure the fix works?

**Review by 4lejandrito — CHANGES_REQUESTED** [review]

File: modules/apps/analytics/analytics-settings-rest-impl/src/main/java/com/liferay/analytics/settings/rest/internal/resource/v1_0/BaseChannelResourceImpl.java [hunk 2]
```diff
@@ -342,81 +342,115 @@ private Collection<Permission> _getPermissions(
 
 		Map<String, Permission> permissions = new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/analytics/analytics-settings-rest-impl/src/main/java/com/liferay/analytics/settings/rest/internal/resource/v1_0/BaseChannelResourceImpl.java:344]: Make sense @4lejandrito!! I agree with, I thought that was checked :sweat_smile:

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl (hunk 1 above)
Comment (4lejandrito) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:982]: Make sure you autogenerate a test that fails without this fix. In a separate PR.

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 3]
```diff
@@ -975,68 +979,114 @@ public abstract class Base${schemaName}ResourceImpl
...
-		}
+			Set<ResourcePermission> resourcePermissionSet;
+
+			for (ResourceAction resourceAction : resourceActions) {
+				actionIds.add(resourceAction.getActionId());
+			}
+
+			resourcePermissionSet = _getResourcePermissionSet(companyId, resourceId, resourceName);
+
+			for (ResourcePermission resourcePermission : resourcePermissionSet) {
```
Comment (4lejandrito) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:1001]: Refactor this loop as discussed. Following the same logic as edit_permissions.jsp.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 4]
```diff
@@ -975,68 +979,114 @@ public abstract class Base${schemaName}ResourceImpl
...
+					newPermission.setActionIds(actionIdsSet.toArray(new String [0]));
+				}
 
-						Collections.addAll(actionsIdsSet, permission.getActionIds());
+					permissions.put(role.getName(), newPermission);
+				}
+			}
 
-						Permission newPermission = _toPermission(resourceActions, resourcePermission, role);
+			if (roleNames !=null) {
```
[the remark this answers is not in the document]
Comment (4lejandrito) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:1025]: You'll get this for free in the refactor I sent you privately.

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 5]
```diff
@@ -975,68 +979,114 @@ public abstract class Base${schemaName}ResourceImpl
...
-			if ((actionIds & bitwiseValue) == bitwiseValue) {
-				actionsIdsSet.add(resourceAction.getActionId());
-			}
-		}
+		resourcePermissionSet = new HashSet<>(resourcePermissions);
+
+		resourcePermissionSet.addAll(
+			resourcePermissionLocalService.getResourcePermissions(
+				companyId, resourceName, ResourceConstants.SCOPE_GROUP_TEMPLATE,
+				String.valueOf(0)));
```
[the remark this answers is not in the document]
Comment (4lejandrito) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl]: ☝️

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 6]
```diff
@@ -975,68 +979,114 @@ public abstract class Base${schemaName}ResourceImpl
...
-		try {
-			Method method = ResourcePermissionLocalService.class.getMethod("getResourcePermissions", String.class);
+			List<String> actionIds = new ArrayList<>();
 
-			resourcePermissions = (List<ResourcePermission>)method.invoke(resourcePermissionLocalService, resourceName);
-		}
-		catch (NoSuchMethodException noSuchMethodException) {
-			resourcePermissions = resourcePermissionLocalService.getResourcePermissions(companyId, resourceName, ResourceConstants.SCOPE_INDIVIDUAL, String.valueOf(resourceId));
-		}
+			Set<ResourcePermission> resourcePermissionSet;
```
[the remark this answers is not in the document]
Comment (4lejandrito) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl]: ☝️

**Review by 4lejandrito — CHANGES_REQUESTED** [review]
See the comments from yesterday.

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl [hunk 7]
```diff
@@ -975,69 +979,82 @@ public abstract class Base${schemaName}ResourceImpl
...
-			Method method = ResourcePermissionLocalService.class.getMethod("getResourcePermissions", String.class);
+		Set<ResourcePermission> resourcePermissionSet = _getResourcePermissionSet(
+			companyId, resourceId, resourceName);
 
-			resourcePermissions = (List<ResourcePermission>)method.invoke(resourcePermissionLocalService, resourceName);
-		}
-		catch (NoSuchMethodException noSuchMethodException) {
-			resourcePermissions = resourcePermissionLocalService.getResourcePermissions(companyId, resourceName, ResourceConstants.SCOPE_INDIVIDUAL, String.valueOf(resourceId));
-		}
+		List<Resource> resources = _getResourceList(resourcePermissionSet);
```
Comment (4lejandrito) [inline modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:995]: call this `getResources` and the same with the variable inside that method.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]

**Comment by sergiojimcos** [comment]
ci:stop

File: modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl (hunk 7 above)
Comment (sergiojimcos) [reply modules/util/portal-tools-rest-builder/src/main/resources/com/liferay/portal/tools/rest/builder/dependencies/base_resource_impl.ftl:995]: I will also update some other method names, to maintain the consistency

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Review by 4lejandrito — APPROVED** [review]

**Comment by yanan** [comment]
ci:test:relevant

**Comment by yanan** [comment]
No unique failures related to the pr, ok to be forwarded.

**Comment by sergiojimcos** [comment]
Sent to Brian manually: https://github.com/brianchandotcom/liferay-portal/pull/132194

**Pull request closed without merging**